consensus/src/consensus/inner.rs

```diff
@@ -0,0 +1,508 @@
+        let transaction_ids: Vec<[u8; 32]> = block.transactions.iter().map(|x| x.id).collect();
+        let (merkle_root, pedersen_merkle_root, _) = txids_to_roots(&transaction_ids);
+
+        let canon = self.storage.canon().await?;
+        // Verify the block header
+        if block.header.previous_block_hash != canon.hash {
+            return Err(anyhow!("attempted to commit a block that wasn't a direct child of tip of canon").into());
+        }
+        let parent_header = self.storage.get_block_header(&canon.hash).await?;
+        if let Err(err) =
+            self.public
+                .parameters
+                .verify_header(&block.header, &parent_header, &merkle_root, &pedersen_merkle_root)
+        {
+            error!("block header failed to verify: {:?}", err);
+            return Ok(false);
+        }
+
+        // Verify block amounts and check that there is a single coinbase transaction
+        let mut coinbase_transaction_count: i32 = 0;
+        let mut total_value_balance = AleoAmount::ZERO;
+
+        for transaction in block.transactions.iter() {
+            let value_balance = transaction.value_balance;
+
+            if value_balance.is_negative() {
+                coinbase_transaction_count += 1;
+            }
+
+            total_value_balance = total_value_balance.add(value_balance);
+        }
+        // Check that there is only 1 coinbase transaction
+        if coinbase_transaction_count > 1 {
+            error!("multiple coinbase transactions");
+            return Ok(false);
+        }
+        // Check that the block value balances are correct
+        let expected_block_reward = crate::get_block_reward(canon.block_height as u32).0;
+        if total_value_balance.0 + expected_block_reward != 0 {
+            trace!("total_value_balance: {:?}", total_value_balance);
+            trace!("expected_block_reward: {:?}", expected_block_reward);
+
+            return Ok(false);
+        }
+
+        // Check that all the transaction proofs verify
+        self.verify_transactions(&block.transactions[..])
+    }
+
+    /// Receive a block from an external source and process it based on ledger state.
+    async fn receive_block(&mut self, block: &SerialBlock) -> Result<(), ConsensusError> {
+        let hash = block.header.hash();
+        match self.storage.get_block_state(&hash).await? {
+            BlockStatus::Unknown => (),
+            BlockStatus::Committed(_) | BlockStatus::Uncommitted => {
+                debug!("Received a pre-existing block");
+                metrics::increment_counter!(DUPLICATE_BLOCKS);
+                return Err(ConsensusError::PreExistingBlock);
+            }
+        }
+        self.storage.insert_block(block).await?;
+
+        self.try_commit_block(&hash, block).await?;
+
+        self.catch_up().await?;
+
+        Ok(())
+    }
+
+    async fn catch_up(&mut self) -> Result<(), ConsensusError> {
```

Comment:
The phrase "catch up" is often used to describe the block sync process, and I initially thought this was it - perhaps "try_to_fast_forward" would be more suitable?